// reactotron-core-contract/dist/commonjs/command.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommandType = void 0;
const CommandType = exports.CommandType = {
  ApiResponse: "api.response",
  AsyncStorageMutation: "asyncStorage.mutation",
  Benchmark: "benchmark.report",
  ClientIntro: "client.intro",
  Display: "display",
  Image: "image",
  Log: "log",
  SagaTaskComplete: "saga.task.complete",
  StateActionComplete: "state.action.complete",
  StateKeysResponse: "state.keys.response",
  StateValuesChange: "state.values.change",
  StateValuesResponse: "state.values.response",
  StateBackupResponse: "state.backup.response",
  StateBackupRequest: "state.backup.request",
  StateRestoreRequest: "state.restore.request",
  StateActionDispatch: "state.action.dispatch",
  StateValuesSubscribe: "state.values.subscribe",
  StateKeysRequest: "state.keys.request",
  StateValuesRequest: "state.values.request",
  CustomCommandRegister: "customCommand.register",
  CustomCommandUnregister: "customCommand.unregister",
  Clear: "clear",
  ReplLsResponse: "repl.ls.response",
  ReplExecuteResponse: "repl.execute.response",
  // these technically are commands only in reactotron-react-native, but I felt lazy so they can live here
  DevtoolsOpen: "devtools.open",
  DevtoolsReload: "devtools.reload",
  EditorOpen: "editor.open",
  Storybook: "storybook",
  Overlay: "overlay"
};
