// reactotron-react-native/dist/commonjs/plugins/overlay/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = OverlayCreator;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _mitt = _interopRequireDefault(require("mitt"));
var _overlay = _interopRequireDefault(require("./overlay"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function OverlayCreator() {
  return function overlay() {
    const emitter = (0, _mitt.default)();
    return {
      /**
       * Fires when any Reactotron message arrives.
       */
      onCommand: command => {
        if (command.type !== "overlay") return;
        // relay this payload on to the emitter
        emitter.emit("overlay", command.payload);
      },
      features: {
        overlay: WrappedComponent => (props = {}) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_reactNative.View, {
          style: {
            flex: 1
          },
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(WrappedComponent, {
            ...props
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_overlay.default, {
            emitter: emitter
          })]
        })
      }
    };
  };
}
