// reactotron-react-native/dist/commonjs/plugins/storybook/storybook.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
class StorybookSwitcher extends _react.Component {
  /**
   * Creates an instance of FullScreenOverlay.
   *
   * @memberOf FullScreenOverlay
   */
  constructor(props) {
    super(props);
    this.state = {
      showStorybook: false
    };

    // when the server sends stuff
    props.emitter.on("storybook", payload => {
      this.setState({
        showStorybook: payload
      });
    });
  }

  /**
   * Draw.
   */
  render() {
    const {
      showStorybook
    } = this.state;
    const {
      storybookUi: StorybookUi,
      children
    } = this.props;
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
      style: {
        flex: 1
      },
      children: showStorybook ? /*#__PURE__*/(0, _jsxRuntime.jsx)(StorybookUi, {}) : children
    });
  }
}
var _default = exports.default = StorybookSwitcher;
