// reactotron-react-native/dist/commonjs/helpers/getReactNativePlatformConstants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getReactNativePlatformConstants;
var _reactNative = require("react-native");
function getReactNativePlatformConstants() {
  const defaults = {
    osRelease: "",
    model: "",
    serverHost: "",
    uiMode: "",
    serial: "",
    forceTouch: false,
    interfaceIdiom: "",
    systemName: ""
  };
  if (_reactNative.Platform.OS === "android") {
    const constants = _reactNative.Platform.constants;
    return {
      ...defaults,
      osRelease: constants.Release,
      model: constants.Model,
      serverHost: constants.ServerHost,
      uiMode: constants.uiMode,
      serial: constants.Serial
    };
  } else if (_reactNative.Platform.OS === "ios") {
    const constants = _reactNative.Platform.constants;
    return {
      ...defaults,
      forceTouch: constants.forceTouchAvailable || false,
      interfaceIdiom: constants.interfaceIdiom,
      systemName: constants.systemName
    };
  }
  return defaults;
}
